File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D22.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day
import java.util.PriorityQueue

class Y15D22(input: String) : Day {
    // Process input
    private val bossHP = input.dropWhile { !it.isDigit() }.takeWhile { it.isDigit() }.toInt()
    private val damage = input.takeLastWhile { it.isDigit() }.toInt()

    enum class Spell(val mana: Int, val duration: Int, val effect: Int) {
        MagicMissile(53, 1, 4),
        Drain(73, 1, 2),
        Shield(113, 6, 7),
        Poison(173, 6, 3),
        Recharge(229, 5, 101)
    }

    data class State(
        val playerHP: Int,
        val bossHP: Int,
        val currentMana: Int,
        val manaSpent: Int,
        val shield: Int,
        val poison: Int,
        val recharge: Int
    ): Comparable<State> {
        val availableMana = currentMana + (if (recharge > 0) Spell.Recharge.effect else 0)

        fun turn(spell: Spell, damage: Int, constantDrain: Int): State {
            // If player dies from constant drain, boss does not sustain any damage, so return early
            if (playerHP - constantDrain == 0) return copy(playerHP = 0)
            var newPlayerHP = playerHP +
                    (if (spell == Spell.Drain) Spell.Drain.effect else 0) -  
                    constantDrain 
            var newBossHP = bossHP -
                    (if (poison > 0) Spell.Poison.effect else 0) - 
                    (if (spell == Spell.MagicMissile) Spell.MagicMissile.effect else 0) -
                    if (spell == Spell.Drain) Spell.Drain.effect else 0
            var newCurrentMana = currentMana -
                    spell.mana +
                    if (recharge > 0) Spell.Recharge.effect else 0
            val newManaSpent = manaSpent + spell.mana
            val newShield = if (spell == Spell.Shield) spell.duration else shield - 1
            val newPoison = if (spell == Spell.Poison) spell.duration else poison - 1
            val newRecharge = if (spell == Spell.Recharge) spell.duration else recharge - 1

            if (newPlayerHP > 0) {
                val armor = if (newShield > 0) Spell.Shield.effect else 0
                newPlayerHP -= (damage - armor).coerceAtLeast(1)
                newBossHP -= if (newPoison > 0) Spell.Poison.effect else 0
                newCurrentMana += if (newRecharge > 0) Spell.Recharge.effect else 0
            }

            return State(
                newPlayerHP, newBossHP, newCurrentMana, newManaSpent,
                newShield - 1, newPoison - 1, newRecharge - 1,
            )
        }
        override fun compareTo(other: State) = other.manaSpent - manaSpent
    }

    /**
     * Dijkstra implementation, maintaining game state and trying all available spells, trying minimum mana states
     * first. 
     */
    private fun solve(constantDrain: Int): Int {
        val states = PriorityQueue<State>()
        states.add(State(50, bossHP, 500, 0, 0, 0, 0))

        var lowestManaWin = Int.MAX_VALUE

        while (states.isNotEmpty()) {
            val current = states.poll()
            
            // try every castable spell and calculate the state
            val nextStates = Spell.entries
                .filter { spell ->
                    if (current.availableMana < spell.mana) {
                        false // not enough mana to cast!
                    } else {
                        // You can only cast if the effect is expired/expiring
                        when (spell) {
                            Spell.Shield -> current.shield <= 1
                            Spell.Poison -> current.poison <= 1
                            Spell.Recharge -> current.recharge <= 1
                            else -> true
                        }
                    }
                }.map {
                    // simulate turn
                    current.turn(it, damage, constantDrain)
                }.filter {
                    when {
                        // boss is dead, so check to see if the mana is less than current lowest. Don't propagate state.
                        it.bossHP <= 0 -> {
                            lowestManaWin = minOf(lowestManaWin, it.manaSpent)
                            false
                        }

                        // either player is dead, player can't cast a spell the next round, or the total mana spent
                        // is higher than the current lowest mana win. In all cases, player meets with failure so
                        // do not propagate state.
                        it.playerHP <= 0 || it.availableMana < 53 || it.manaSpent > lowestManaWin -> false
                        
                        // otherwise, not finished so propagate state
                        else -> true
                    }
                }
            states.addAll(nextStates)
        }
        return lowestManaWin
    }

    override fun part1() = solve(0)
    override fun part2() = solve(1)
}

fun main() = Day.runDay(Y15D22::class)

//    Class creation: 2ms
//    Part 1: 1824 (222ms)
//    Part 2: 1937 (25ms)
//    Total time: 250ms

File: rust/advent/src/bin/2015/y15d22.rs
use std::{cmp::max, collections::BinaryHeap};

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input = (Int, Int);
type Output = Int;
type Int = i32;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Spell {
    MagicMissile,
    Drain,
    Shield,
    Poison,
    Recharge,
}

impl Spell {
    fn mana(&self) -> Int {
        match self {
            Spell::MagicMissile => 53,
            Spell::Drain => 73,
            Spell::Shield => 113,
            Spell::Poison => 173,
            Spell::Recharge => 229,
        }
    }

    fn duration(&self) -> Int {
        match self {
            Spell::MagicMissile => 1,
            Spell::Drain => 1,
            Spell::Shield => 6,
            Spell::Poison => 6,
            Spell::Recharge => 5,
        }
    }

    fn effect(&self) -> Int {
        match self {
            Spell::MagicMissile => 4,
            Spell::Drain => 2,
            Spell::Shield => 7,
            Spell::Poison => 3,
            Spell::Recharge => 101,
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Hash)]
struct State {
    player_hp: Int,
    boss_hp: Int,
    current_mana: Int,
    mana_spent: Int,
    shield: Int,
    poison: Int,
    recharge: Int,
}

impl State {
    fn available_mana(&self) -> Int {
        self.current_mana + if self.recharge > 0 { Spell::Recharge.effect() } else { 0 }
    }

    fn turn(&self, spell: Spell, damage: Int, constant_drain: Int) -> Self {
        // If player dies from constant drain, boss does not sustain any damage, so return early
        if self.player_hp - constant_drain == 0 {
            return Self { player_hp: 0, ..*self }
        }

        let mut new_player_hp = self.player_hp +
            (if spell == Spell::Drain { Spell::Drain.effect() } else { 0 }) -
            constant_drain;
        let mut new_boss_hp = self.boss_hp -
            if self.poison > 0 { Spell::Poison.effect() } else { 0 } -
            if spell == Spell::MagicMissile { spell.effect() } else { 0 } -
            if spell == Spell::Drain { spell.effect() } else { 0 };
        let mut new_current_mana = self.current_mana - spell.mana() +
            if self.recharge > 0 { Spell::Recharge.effect() } else { 0 };
        let new_mana_spent = self.mana_spent + spell.mana();
        let new_shield = if spell == Spell::Shield { spell.duration() } else { self.shield - 1 };
        let new_poison = if spell == Spell::Poison { spell.duration() } else { self.poison - 1 };
        let new_recharge = if spell == Spell::Recharge { spell.duration() } else { self.recharge - 1 };

        if new_player_hp > 0 {
            let armor = if new_shield > 0 { Spell::Shield.effect() } else { 0 };
            new_player_hp -= max(damage - armor, 1);
            new_boss_hp -= if new_poison > 0 { Spell::Poison.effect() } else { 0 };
            new_current_mana += if new_recharge > 0 { Spell::Recharge.effect() } else { 0 };
        }

        Self {
            player_hp: new_player_hp,
            boss_hp: new_boss_hp,
            current_mana: new_current_mana,
            mana_spent: new_mana_spent,
            shield: new_shield - 1,
            poison: new_poison - 1,
            recharge: new_recharge - 1,
        }
    }
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.mana_spent.cmp(&other.mana_spent)
    }
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 22).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(input), stopwatch.lap().report());
    println!("2. {} ({})", part2(input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.get_numbers().collect_tuple().unwrap()
}

fn solve((boss_hp, boss_damage): Input, constant_drain: Int) -> Output {
    let mut q = BinaryHeap::new();
    q.push(State { player_hp: 50, boss_hp, current_mana: 500, mana_spent: 0, shield: 0, poison: 0, recharge: 0 });

    let mut lowest_mana_win = Int::MAX;

    while let Some(current) = q.pop() {
        let next_states = [Spell::MagicMissile, Spell::Drain, Spell::Shield, Spell::Poison, Spell::Recharge]
            .into_iter()
            .filter(|spell| {
                if current.available_mana() < spell.mana() {
                    false
                } else {
                    match spell {
                        Spell::Shield => current.shield <= 1,
                        Spell::Poison => current.poison <= 1,
                        Spell::Recharge => current.recharge <= 1,
                        _ => true
                    }
                }
            })
            .map(|spell| {
                current.turn(spell, boss_damage, constant_drain)
            })
            .filter(|state| {
                if state.boss_hp <= 0 {
                    if state.mana_spent < lowest_mana_win {
                        lowest_mana_win = state.mana_spent;
                    }
                    false
                } else if state.player_hp <= 0 || state.available_mana() < 53 || state.mana_spent > lowest_mana_win {
                    false
                } else {
                    true
                }
            })
            .collect_vec();
        for next in next_states {
            q.push(next);
        }
    }
    lowest_mana_win
}

fn part1(input: Input) -> Output {
    solve(input, 0)
}

fn part2(input: Input) -> Output {
    solve(input, 1)
}

#[test]
fn default() {
    let input = get_input(15, 22).unwrap();
    let input = parse_input(&input);
    assert_eq!(1824, part1(input));
    assert_eq!(1937, part2(input));
}

// Input parsed (18μs)
// 1. 1824 (98ms)
// 2. 1937 (5ms)
// Total: 104ms